Custom delimiters

Swieckowski opened the pull request:
<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines;
https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first three fields are mandatory:
-->

**- Summary**

<!--
Explain the **motivation** for making this change.
What existing problem does the pull request solve?
-->
This allows users to specify their own custom delimiters for frontmatter if they've specified a format type. Closes #980 

**- Test plan**

<!--
Demonstrate the code is solid.
Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.
-->
I checked all the different types of format settings with unique delimiters. Not sure if I could have done more.

**- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**- A picture of a cute animal (not mandatory but encouraged)**

verythorough commented:
Deploy preview for *netlify-cms-www* ready!

Built with commit 50a5ed03f39cc9efd242068f80fc06609f6e9e48

https://deploy-preview-1064--netlify-cms-www.netlify.com

verythorough commented:
Deploy preview for *cms-demo* ready!

Built with commit 50a5ed03f39cc9efd242068f80fc06609f6e9e48

https://deploy-preview-1064--cms-demo.netlify.com

erquhart commented:
Thanks @Swieckowski! Going to take a look at this soon.

Swieckowski commented:
Good to hear @erquhart, hopefully I can do a PR on my other issue claim soon enough as well.

tech4him1 commented:
I'll review ASAP.

tech4him1 commented on src/formats/formats.js:
```diff
@@ -37,40 +37,42 @@ export function formatByExtension(extension) {
-    'yaml-frontmatter': FrontmatterYAML,
+    'json-frontmatter': frontmatterJSON(customDelimiter),
+    'toml-frontmatter': frontmatterTOML(customDelimiter),
+    'yaml-frontmatter': frontmatterYAML(customDelimiter),
```
I do wonder if there is a better way to avoid the overhead that is incurred here. The benefit of the single objects that we had before was that there was only a single instance of each created for the entire CMS (7 total instances). Now, we're creating three new objects for each *page*, then, throwing them all away (unless one is used). That just seems a little wasteful to me.

To pass a custom parameter, we're obviously going to have to do more than one instance. I think the cleanest, most sustainable way would be to create only one instance per *collection*.
For now, I wonder if we could just wait to pass the `customDelimiter` in until after `formatByName` returns the function. I haven't thought this through that much, though.

@erquhart Do you think this could cause performance problems now, or would you like to merge this, and change it later? I haven't worked with JS enough to know how expensive this really is.

erquhart replied:
I doubt any real performance penalty would happen here. I did have the same thoughts on improvements, but trying to prioritize timely merges over ideal code. This passed the "good enough" test for me, personally.

Swieckowski replied:
So because I'm passing in customDelimiters here, you're saying that each object gets actually instantiated while before it did not? That totally escaped me, I can test out a few things and make this more optimized if you'd like.

tech4him1 replied:
@Swieckowski I think, like Shawn said, that it's probably not worth delaying the PR just to try to optimize the code. The difference is that a new object gets created with each function call, instead of just reusing the same one. I don't really want to delay it just for that, and we can optimize later if necessary. Do you agree?

Swieckowski replied:
Sure! I was just trying to make sure I was following your logic correctly.

tech4him1 commented on src/formats/formats.js:
```diff
@@ -37,40 +37,42 @@ export function formatByExtension(extension) {
   if (filePath) {
     const fileExtension = filePath.split('.').pop();
-    return formatByExtension(fileExtension);
+    return formatByExtension(fileExtension, customDelimiter);
```
I don't think we need to pass `customDelimiter` to `formatByExtension()` when it doesn't use it. That can always be added later if needed.

Swieckowski replied:
I just removed the unnecessary passing of customDelimiter to formatByExtension(), it was a remnant of a different train of thought. Sorry I didn't have it cleaned up before.  

Are there any other steps I should take with this?

tech4him1 replied:
Nope, thank you.

tech4him1 commented on src/formats/formats.js:
```diff
@@ -37,40 +37,42 @@ export function formatByExtension(extension) {
 
 export function resolveFormat(collectionOrEntity, entry) {
+  // Check for custom delimiter
+  const customDelimiter = collectionOrEntity.get('frontmatter_delimiter');
```
I think we need to throw if a `frontmatter_delimiter` is set without an explicit `format` being set. This might be better checked in the reducer config validation section: https://github.com/netlify/netlify-cms/blob/e9ec6def04d8a10bcf28173780d9939b30bc195e/src/reducers/collections.js#L29-L41

tech4him1 replied:
@Swieckowski I think we discussed this before, was there something that I missed?

Swieckowski replied:
Ah yes, I remember bringing that up myself but don't recall a specific response, sorry. Should I get on that asap?

tech4him1 replied:
No problem! I'd like to have it in before merge -- if you're busy, I'd be glad to do it.

Swieckowski replied:
I can have it done by 2 or 3pm tomorrow if that's alright.

tech4him1 replied:
Oh, no hurry at all, thank you!

Swieckowski replied:
@tech4him1 I've been a bit swamped with job search stuff haha, but I hope my latest commits address what you were talking about. We now throw an error if someone has a frontmatter_delimiter set without setting their format or not setting it to the proper frontmatter formats.

Swieckowski replied:
Also, since I've added tests and documentation, how do I add that to my contribution? The add contributor script is failing on me, perhaps because I've already run it once on my github username?

erquhart commented:
Eww resolving conflicts in GitHub's UI results in a merge commit 🤦‍♂️ .

Anywho, feel free to merge when you're happy with this @tech4him1. Thanks again @Swieckowski!

verythorough commented on website/site/content/docs/configuration-options.md:
```diff
@@ -100,7 +100,9 @@ You may also specify a custom `extension` not included in the list above, as lon
-The explicit `yaml-frontmatter`, `toml-frontmatter`, and `json-frontmatter` formats above do not currently support custom delimiters. We use `---` for YAML, `+++` for TOML, and `{` `}` for JSON. If a file has frontmatter inside other delimiters it will be included as part of the body text.
+The explicit `yaml-frontmatter`, `toml-frontmatter`, and `json-frontmatter` formats support custom delimiters. We use `---` for YAML, `+++` for TOML, and `{` `}` for JSON by default, but you can configure a custom delimiter.
+You can configure a custom delimiter with `frontmatter_delimiter`, but it only applies if you have a frontmatter format specified.
+If a file has frontmatter inside other delimiters it will be included as part of the body text.
```
Thanks for including this change in the docs!

Could you format this like the other config options? Something like:

> - `frontmatter_delimiter`:  *description here*

I think you could also add the default delimiters (`+++` etc) to the descriptions for the explicit frontmatter formats instead of putting them all together in the `frontmatter_delimiter` description.

Swieckowski replied:
@verythorough I took your comments into consideration, I hope the changes reflect the direction you were looking for.

tech4him1 commented on src/reducers/collections.js:
```diff
@@ -38,6 +38,12 @@ function validateCollection(configCollection) {
     // Cannot infer format from extension.
     throw new Error(`Please set a format for collection "${ collectionName }". Supported formats are ${ supportedFormats.join(',') }`);
   }
+  const frontmatterFormats = ['yaml-frontmatter','toml-frontmatter','json-frontmatter']
```
I'd rather have this list imported from `src/formats/formats.js`, like we did for the `supportedFormats` one above. Just easier to keep it together.

Swieckowski replied:
Fixed. I will think more carefully about where I place things from now, thanks for your patience.

tech4him1 replied:
Oh no, not a problem at all. Sorry to bug you with little things.

tech4him1 commented:
@Swieckowski Looks like the rebase messed up the all contributors, you should be abel to run it again and add tests and docs.

Swieckowski commented:
@tech4him1 I think I've addressed all of your concerns except for the minor optimization. Which I could slightly ameliorate with a switch statement refactor instead of the object property approach currently used, but I don't think the derailment from the norm is worth the slight optimization (which wouldn't bring the performance back to the same level anyway). 

The community's been really helpful, and I think my next PR will benefit from all the input I've received.
